internal/action/command.go

```diff
@@ -361,6 +361,10 @@ func reloadRuntime(reloadPlugins bool) {
 		parsedSettings := config.ParsedSettings()
 		defaultSettings := config.DefaultAllSettings()
 		for k := range defaultSettings {
+			if k == "fileformat" {
+				// reload should not override auto-detected file format
+				continue
```

Comment:
Doesn't work for me: `reload` still overrides the auto-detected format.

Looks like the reason is, again, that setting `filetype` (by `DoSetOptionNative()` below) triggers setting other options, including `fileformat`.

Actually now it seems to me that this hardcoded skipping of `fileformat` here is not quite correct anyway. In some cases (namely, if the file is empty) we don't autodetect `fileformat` in `NewBuffer()`, and in such case we probably do want `reload` to update `fileformat`.

So it seems the right way is to add `fileformat` to `b.LocalSettings` in `NewBuffer()` (in the case when we do autodetect it)? And that will address this issue with `DoSetOptionNative("filetype")` as well (since it skips `LocalSettings`).

Reply:
Actually, besides this issue, it seems pointless to let `DoSetOptionNative("filetype")` call `InitLocalSettings()` and update other options when it is called from `reload`, since in `reload` we call `InitLocalSettings()` and update all other options anyway (so as a result, we needlessly update all local settings twice).

So it seems we should move this `filetype` logic (calling `InitLocalSettings()` + updating other settings) from `DoSetOptionNative()` to `SetOptionNative()`, as a special case?

Reply:
> So it seems the right way is to add `fileformat` to `b.LocalSettings` in `NewBuffer()` (in the case when we do autodetect it)? And that will address this issue with `DoSetOptionNative("filetype")` as well (since it skips `LocalSettings`).

Yes, this is a good idea and already evaluated with success.

> So it seems we should move this `filetype` logic (calling `InitLocalSettings()` + updating other settings) from `DoSetOptionNative()` to `SetOptionNative()`, as a special case?

Maybe easier and without the "special case" is to skip entries in `config.LocalSettings` (was already present in one version but lost during the various reworks) from being forwarded into `DoSetOptionNative()` upon `reload`?
Why should the `reload` touch common local-only options in anyway, especially `filetype` with `unknown` and starting a further autodetection? This is done anyway, in case the `filetype` is `unknown` from the beginning due to the following call of `UpdateRules()`.
So do we need to reset `filetype` to restart the autodetection again? I don't think so and new rules will be applied anyway too.

Reply:
Agree, skipping local-only options here seems to be exactly what we should do.

Or not exactly... `config.LocalSettings` is not *exactly* the same thing as local-only settings. Micro actually allows settings a default global value of `filetype` in `settings.json`, e.g. `"filetype": "c"`, and this value is then set for any buffer for which micro doesn't autodetect the file type. So we probably want `reload` to reload `filetype` in this case (but make sure that it doesn't override autodetected filetypes)? 

So... you can guess what I'm gonna suggest?

Reply:
I can guess, but I might be wrong.

Move everything out of `DoSetOptionNative("filetype")` into a own exported function e.g. `ReloadSettings()` and call this for the `filetype` and the `reload` case.
By this we should get rid of the multiple reset cases and don't duplicate the same code necessary.

Reply:
I meant something different: add `fileformat` to `b.LocalSettings` when we override it in `UpdateRules()`, and do what you suggested ("skip entries in `config.LocalSettings` from being forwarded into `DoSetOptionNative()` upon `reload`") but not with `config.LocalSettings` but with `b.LocalSettings`.

Unless there are some reasons not to do that (I haven't thought that out carefully)?

Reply:
Ok, so my first impression was right, but I got confused by the strikethroughs.
I will see how it reacts when I add `filetype` to `b.LocalSettings` when it is set in `UpdateRules()`.

Due to https://github.com/zyedidia/micro/pull/3343#discussion_r1710450982 we need to properly update that section anyway.

Reply:
Ah, those strikethroughs are there just because after I wrote that comment, I realized that setting `filetype` in `settings.json` completely disables filetype autodetection. But this fact doesn't seem to change much for us here.

Reply:
From workflow it does seem to make sense setting `b.LocalSettings["filetype"]` to `true` only in case `foundDef` was set too, otherwise `reload` wouldn't set a possibly added global `filetype` when **no** buffer-local one was forced by the user. In case it was forced by the user, which is possible every time, it isn't overwritten by `reload` then.

Does this fit to your expectation as well?

Reply:
`foundDef` seems to play quite a different role: it is to distinguish if the found syntax file is a built-in one or a user-defined one (regardless of whether the filetype is autodetected, globally predefined or locally forced by the user)?

The distinction we are interested in rather lies along the lines of `ft == "unknown" || ft == ""`?

Reply:
Ah yes, I can remember...long time ago.
Anyway, in case something was found (where ever) it ends in `b.SyntaxDef != nil` and setting it there doesn't really help nor does it lead to comprehensible behavior.

I suggest to go with the following behavior:
Unless it isn't forced by the user via setting the `filetype` it's possible to overwrite it with reload. It fits to the behavior that it applies the (global) `filetype` read from the `settings.json` on openening `micro` to the opened buffer anyway.

Reply:
Maybe... It's hard to discuss an implementation that doesn't exist yet (since the question how to address this issue depends on how will we address other, closely related issues - [this one](https://github.com/zyedidia/micro/pull/3343#discussion_r1710336694) and [this one](https://github.com/zyedidia/micro/pull/3343#discussion_r1710088161)).